modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java

```diff
@@ -0,0 +1,522 @@
+        var raftServers = new HashMap<NetworkAddress, JraftServerImpl>();
+        int nodes = 3;
+        Predicate<NetworkAddress> isServerAddress = addr -> true;
+
+        TopologyAwareRaftGroupService raftClient = startCluster(
+                testInfo,
+                clusterServices,
+                raftServers,
+                isServerAddress,
+                nodes,
+                PORT_BASE
+        );
+
+        raftClient.refreshLeader().get();
+
+        var clientClusterService = clusterService(testInfo, PORT_BASE + nodes + 1, new StaticNodeFinder(getNetworkAddresses(nodes)));
+        clientClusterService.start();
+
+        TopologyAwareRaftGroupService raftClientNoInitialNotify = startTopologyAwareClient(
+                clientClusterService,
+                clusterServices,
+                isServerAddress,
+                nodes,
+                null,
+                false
+        );
+
+        AtomicReference<ClusterNode> leaderRef = new AtomicReference<>();
+        AtomicReference<ClusterNode> leaderRefNoInitialNotify = new AtomicReference<>();
+        AtomicInteger callsCount = new AtomicInteger();
+
+        subscribeLeader(raftClient, (node, term) -> leaderRef.set(node), "New leader: {}");
+
+        for (int i = 0; i < 2; i++) {
+            unsubscribeLeader(raftClientNoInitialNotify);
+
+            subscribeLeader(raftClientNoInitialNotify, (node, term) -> {
+                callsCount.incrementAndGet();
+                leaderRefNoInitialNotify.set(node);
+            }, "New leader (client without initial notification): {}");
+        }
+
+        assertTrue(callsCount.get() <= 1);
+
+        assertTrue(waitForCondition(() -> leaderRef.get() != null, 10_000));
+
+        ClusterNode leader = leaderRef.get();
+
+        assertNotNull(leader);
+
+        log.info("Leader: " + leader);
+
+        afterClusterInit(leader.name());
+
+        var raftServiceToStop = raftServers.remove(new NetworkAddress("localhost", leader.address().port()));
+        raftServiceToStop.stopRaftNodes(GROUP_ID);
+        raftServiceToStop.stop();
+
+        afterNodeStop(leader.name());
+
+        clusterServices.remove(new NetworkAddress("localhost", leader.address().port())).stop();
+
+        assertTrue(waitForCondition(() -> !leader.equals(leaderRef.get()), 10_000));
+        assertTrue(waitForCondition(() -> !leader.equals(leaderRefNoInitialNotify.get()), 1000));
+
+        log.info("New Leader: " + leaderRef.get());
+
+        afterLeaderChange(leaderRef.get().name());
+
+        raftClientNoInitialNotify.refreshLeader().get();
+
+        assertEquals(raftClientNoInitialNotify.leader().consistentId(), leaderRef.get().name());
+
+        stopCluster(clusterServices, raftServers, List.of(raftClient, raftClientNoInitialNotify), nodes);
+
+        clientClusterService.stop();
+    }
+
+    @Test
+    public void testChangeLeaderForce(TestInfo testInfo) throws Exception {
```

Comment:
a lot of code here and in `testChangeLeaderWhenActualLeft` is duplicated, like 50 lines in the beginning, please refactor

Reply:
also please add some comments or make code more readable, it is really hard to understand what is going on in these tests

Reply:
moved the cluster initialization to the separate method